src/mem/x86_64.rs

```diff
@@ -0,0 +1,69 @@
+use super::c_int;
+
+// On recent Intel processors, "rep movsb" and "rep stosb" have been enhanced to
+// select the best microarchitectural implementation based on length and
+// alignment. This allows for very fast memcmp/memset implementations.
+// See "Intel® 64 and IA-32 Architectures Optimization Reference Manual"
+//  - Enhanced REP MOVSB and STOSB operation (ERMSB) (section 3.7.7)
+//  - Fast Short REP MOV (FSRM)
+
+#[cfg_attr(all(feature = "mem", not(feature = "mangled-names")), no_mangle)]
+pub unsafe extern "C" fn memcpy(dest: *mut u8, src: *const u8, count: usize) -> *mut u8 {
+    let qword_count = count >> 3;
+    let byte_count = count & 0b111;
+    asm!(
+        "rep movsq [rdi], [rsi]",
+        "mov ecx, {byte_count:e}",
+        "rep movsb [rdi], [rsi]",
+        byte_count = in(reg) byte_count,
+        inout("rcx") qword_count => _,
+        inout("rdi") dest => _,
+        inout("rsi") src => _,
+        options(nostack, preserves_flags)
+    );
+    dest
+}
+
+#[cfg_attr(all(feature = "mem", not(feature = "mangled-names")), no_mangle)]
+pub unsafe extern "C" fn memmove(dest: *mut u8, src: *const u8, count: usize) -> *mut u8 {
+    let dest_end = dest.offset(count as isize);
+    let src_end = src.offset(count as isize);
+    if src >= dest as *const u8 || src_end <= dest as *const u8 {
+        return memcpy(dest, src, count);
+    }
+    // copy backwards
+    let qword_count = count >> 3;
+    let byte_count = count & 0b111;
+    asm!(
+        "std",
+        "rep movsq [rdi], [rsi]",
+        "mov ecx, {byte_count:e}",
+        "add rdi, 7",
+        "add rsi, 7",
+        "rep movsb [rdi], [rsi]",
+        "cld",
+        byte_count = in(reg) byte_count,
+        inout("rcx") qword_count => _,
+        inout("rdi") dest_end.sub(8) => _,
+        inout("rsi") src_end.sub(8) => _,
+        options(nostack, preserves_flags)
```

Comment:
`add` clobbers the flags.

Reply:
Also, it clears the directions flag, which could be set before

Reply:
Fixed (due to the add). Note that the direction flag must be cleared before exiting an ASM block, andit doesn't affect `preserves_flags`.